primitives/numeric/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+        fn decode<I: parity_scale_codec::Input>(
+            input: &mut I,
+        ) -> Result<Self, parity_scale_codec::Error> {
+            let NumericScaleHelper { mantissa, scale } = NumericScaleHelper::decode(input)?;
+            match Numeric::try_new(mantissa, scale) {
+                Ok(numeric) => Ok(numeric),
+                Err(NumericError::MantissaTooLarge) => {
+                    Err("error decoding numeric: mantissa too large".into())
+                }
+                Err(NumericError::ScaleTooLarge) => {
+                    Err("error decoding numeric: scale too large".into())
+                }
+                // Not possible to get malformed or negative value from mantissa and scale
+                Err(NumericError::Malformed | NumericError::Negative) => unreachable!(),
+            }
+        }
+    }
+}
+
+mod schema_ {
+    use iroha_schema::{
+        Declaration, Ident, IntoSchema, MetaMap, Metadata, NamedFieldsMeta, TypeId,
+    };
+
+    use super::*;
+
+    impl TypeId for Numeric {
+        fn id() -> Ident {
+            "Numeric".to_string()
+        }
+    }
+
+    impl IntoSchema for Numeric {
+        fn type_name() -> Ident {
+            "Numeric".to_string()
+        }
+
+        fn update_schema_map(metamap: &mut MetaMap) {
+            if !metamap.contains_key::<Self>() {
+                if !metamap.contains_key::<u128>() {
+                    <u128 as iroha_schema::IntoSchema>::update_schema_map(metamap);
+                }
+                if !metamap.contains_key::<u32>() {
+                    <u32 as iroha_schema::IntoSchema>::update_schema_map(metamap);
+                }
+
+                metamap.insert::<Self>(Metadata::Struct(NamedFieldsMeta {
+                    declarations: vec![
+                        Declaration {
+                            name: "mantissa".to_string(),
+                            ty: core::any::TypeId::of::<u128>(),
+                        },
+                        Declaration {
+                            name: "scale".to_string(),
+                            ty: core::any::TypeId::of::<u32>(),
+                        },
+                    ],
+                }));
+            }
+        }
+    }
+}
+
+#[cfg(any(feature = "ffi_export", feature = "ffi_import"))]
+mod ffi {
+    //! Manual implementations of FFI related functionality
+    #![allow(unsafe_code)]
+
+    use super::*;
+
+    iroha_ffi::ffi_type! {
+        unsafe impl Transparent for Numeric {
+            type Target = u128;
+        }
+    }
+
+    iroha_ffi::ffi_type! {
+        unsafe impl Transparent for NumericConstrained {
+            type Target = Numeric;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use parity_scale_codec::{Decode, Encode};
+
+    use super::*;
+
+    #[test]
+    fn check_add() {
+        let a = Numeric::new(10, 0);
+        let b = Numeric::new(9, 3);
+
+        assert_eq!(a.checked_add(b), Some(Numeric::new(10009, 3)));
```

Comment:
So, unlike in doubles, a number with larger scale is actually smaller? This is somewhat surprising, but probably makes sense for financial calculations.. Probably worth documenting it in some form, for example by providing a formula, similar to what the underlying docs for `rust_decimal::Decimal` say:

> The finite set of values of type `Decimal` are of the form m / 10<sup>e</sup>, where m is an integer such that -2<sup>96</sup> < m < 2<sup>96</sup>, and e is an integer between 0 and 28 inclusive

Reply:
> So, unlike in doubles, a number with larger scale is actually smaller?

Yes. 
I think it makes sense from the point of view that numbers with positive scale (e.g $1300 = 13 * 10^3$) can be replaced with number with scale 0.


I will add docs about value representation and range.